src/seraphis_wallet/encrypted_file.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include "crypto/chacha.h"
+#include "crypto/crypto.h"
+#include "file_io_utils.h"
+#include "serialization/binary_archive.h"
+#include "serialization/containers.h"
+#include "serialization/crypto.h"
+#include "serialization/serialization.h"
+#include "serialization/string.h"
+#include "storages/portable_storage_template_helper.h"
+#include "string_coding.h"
+
+#include <type_traits>
+
+struct encrypted_file
+{
+    std::string encrypted_data;
+    crypto::chacha_iv iv;
+
+    BEGIN_SERIALIZE_OBJECT()
+    VERSION_FIELD(0)
+    FIELD(encrypted_data)
+    FIELD(iv)
+    END_SERIALIZE()
+};
+
+template <class T> bool read_encrypted_file(std::string path, const crypto::chacha_key &key, T &struct_out)
+{
+    std::string buf;
+    if (!epee::file_io_utils::load_file_to_string(path, buf))
+        return false;
+
+    encrypted_file file;
+
+    binary_archive<false> file_ar{epee::strspan<std::uint8_t>(buf)};
+    if (!::serialization::serialize(file_ar, file))
+        return false;
+
+    std::string decrypted_data;
+    decrypted_data.resize(file.encrypted_data.size());
+    crypto::chacha20(file.encrypted_data.data(), file.encrypted_data.size(), key, file.iv, &decrypted_data[0]);
+
+    binary_archive<false> ar{epee::strspan<std::uint8_t>(decrypted_data)};
```

Comment:
I would consider using the epee binary output. It is more flexible when adding new fields than this cryptonote scheme (which is more compact).

Reply:
> using the epee binary output

Which class is this exactly?

The whole `wallet2` code is pretty much standardized on `binary_archive` now, especially since moneromooo deprecated the use of the outdated Boost serialization a few years back. I think `binary_archive` has proven itself quite well over all that time, and would hesitate to give up the standardization.

Reply:
> Which class is this exactly?


> The whole wallet2 code is pretty much standardized on binary_archive now

I forgot about that. The format is more compact, but you have to remember to add a version field manually in most cases (or some type field), as the field ordering is strict and implied.

The epee binary format is close to msgpack in terms of performance and flexibility (especially when adding new fields).

Reply:
Let us stay with the same stuff that `wallet2` also uses. Using the same serialization system for the file as a whole as for individual classes seems ok to me for the time being.